File: Chess/src/main/java/game/pieces/Knight.java
package game.pieces;

import game.boards.Board;
import game.boards.Tile;
import game.move.Move;
import game.resources.PIECE_Configurations;
import game.util.BoardUtilities;

import java.util.ArrayList;
import java.util.List;


public class Knight extends Piece {

    public Knight(Team team) {
        super(team, PieceTypes.KNIGHT);
    }

    @Override
    public List<Move> availableMoves(Board board, Coordinate currentCoord) {

        List<Move> possibleMoves = new ArrayList<Move>();

        Tile destinationTile;

        for (Coordinate coord : PIECE_Configurations.KNIGHT_MOVES) {

            if (!BoardUtilities.isValidCoordinate(currentCoord.plus(coord))) {
                continue;
            }
            destinationTile = board.getTile(currentCoord.plus(coord));

            if (!destinationTile.hasPiece()) {
                possibleMoves.add(new Move(board, board.getTile(currentCoord), destinationTile));
            } else {
                if (destinationTile.getPiece().getTeam() != this.getTeam()) {
                    possibleMoves.add(new Move(board, board.getTile(currentCoord), destinationTile));
                }
            }
        }
        return possibleMoves;
    }
}

File: Chess/src/main/kotlin/hummel/game/piece/Knight.kt
package main.kotlin.hummel.game.piece

import main.kotlin.hummel.game.Move
import main.kotlin.hummel.game.Pieces
import main.kotlin.hummel.game.board.Board
import main.kotlin.hummel.game.board.Tile
import main.kotlin.hummel.game.util.BoardUtilities

class Knight(team: Team) : Piece(team, PieceTypes.KNIGHT) {
    override fun availableMoves(board: Board, currentCoord: Coordinate): List<Move> {
        val possibleMoves: MutableList<Move> = ArrayList()
        var destinationTile: Tile
        for (coord in Pieces.KNIGHT_MOVES) {
            if (!BoardUtilities.isValidCoordinate(currentCoord.plus(coord))) {
                continue
            }
            destinationTile = board.getTile(currentCoord.plus(coord))
            if (!destinationTile.hasPiece()) {
                possibleMoves.add(Move(board, board.getTile(currentCoord), destinationTile))
            } else {
                if (destinationTile.piece.team !== team) {
                    possibleMoves.add(Move(board, board.getTile(currentCoord), destinationTile))
                }
            }
        }
        return possibleMoves
    }
}